go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -558,3 +570,298 @@ func recalculatePKColsInfoByColumnNames(uniqueKeyColumnNames []string, colInfos
 	}
 	return pkColInfos
 }
+
+// stashSecondaryKeys temporarily DROPs all secondary keys from the table schema
+// and stashes an ALTER TABLE statement that will be used to recreate them at the
+// end of the copy phase.
+func (vr *vreplicator) stashSecondaryKeys(ctx context.Context, tableName string) error {
+	if !vr.supportsDeferredSecondaryKeys() {
+		return fmt.Errorf("deferring secondary key creation is not supported for %s workflows",
+			binlogdatapb.VReplicationWorkflowType_name[vr.WorkflowType])
+	}
+	secondaryKeys, err := vr.getTableSecondaryKeys(ctx, tableName)
+	if err != nil {
+		return err
+	}
+	if len(secondaryKeys) > 0 {
+		alterDrop := &sqlparser.AlterTable{
+			Table: sqlparser.TableName{
+				Qualifier: sqlparser.NewIdentifierCS(vr.dbClient.DBName()),
+				Name:      sqlparser.NewIdentifierCS(tableName),
+			},
+		}
+		alterReAdd := &sqlparser.AlterTable{
+			Table: sqlparser.TableName{
+				Qualifier: sqlparser.NewIdentifierCS(vr.dbClient.DBName()),
+				Name:      sqlparser.NewIdentifierCS(tableName),
+			},
+		}
+		for _, secondaryKey := range secondaryKeys {
+			// should never happen
+			if secondaryKey.Info.Primary {
+				continue
+			}
+			alterDrop.AlterOptions = append(alterDrop.AlterOptions,
+				&sqlparser.DropKey{
+					Name: secondaryKey.Info.Name,
+					Type: sqlparser.NormalKeyType,
+				},
+			)
+			alterReAdd.AlterOptions = append(alterReAdd.AlterOptions,
+				&sqlparser.AddIndexDefinition{
+					IndexDefinition: secondaryKey,
+				},
+			)
+		}
+		action, err := json.Marshal(PostCopyAction{
+			Type: PostCopyActionSQL,
+			Task: sqlparser.String(alterReAdd),
+		})
+		if err != nil {
+			return err
+		}
+		insert, err := sqlparser.ParseAndBind(sqlCreatePostCopyAction, sqltypes.Uint32BindVariable(vr.id),
+			sqltypes.StringBindVariable(tableName), sqltypes.StringBindVariable(string(action)))
+		if err != nil {
+			return err
+		}
+		// Use a new DB client to avoid interfering with open transactions
+		// in the shared client as DDL includes an implied commit.
+		// We're also NOT using a DBA connection here because we want to
+		// be sure that the commit fails if the instance is somehow in
+		// READ-ONLY mode.
+		dbClient := vr.vre.dbClientFactoryFiltered()
+		if err := dbClient.Connect(); err != nil {
+			log.Errorf("unable to connect to the database when saving secondary keys for deferred creation: %v", err)
+			return err
+		}
+		defer dbClient.Close()
+		if _, err := dbClient.ExecuteFetch(insert, 1); err != nil {
+			return err
+		}
+		if _, err := dbClient.ExecuteFetch(sqlparser.String(alterDrop), 1); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (vr *vreplicator) getTableSecondaryKeys(ctx context.Context, tableName string) ([]*sqlparser.IndexDefinition, error) {
```

Comment:
I'm just noting that we drop and re-add all types of secondary keys, including `UNIQUE`s. This is fine, because `vreplication` is unique-compliant and does not need a `UNIQUE` constraint on the target table in order to validate anything.

Reply:
However, if you drop multiple `FULLTEXT` indexes, that's a problem, because you cannot add more than one `FULLTEXT` key in a single `ALTER TABLE` statement.

Reply:
It's implied that the source table must not have any foreign keys. That's already the case today, there is no added restriction, so that's fine.

Reply:
Great points! I wasn't aware of the full text limitations. It would be quite rare, I think, to have more than one in a table but another limitation I realized is that adding one requires a full table rebuild. So I will explicitly leave them in place:
```
		for _, secondaryKey := range secondaryKeys {
			// Primary should never happen. Fulltext keys are
			// not supported for deferral and retained during
			// the copy phase as it has some unique behaviors
			// and constraints:
			// - Adding a fulltext key requires a full table
			//   rebuild to add the internal FTS_DOC_ID field
			//   to each record.
			// - You can not add/remove multiple fulltext keys
			//   in a single ALTER statement.
			if secondaryKey.Info.Primary || secondaryKey.Info.Fulltext {
				continue
			}
```

Reply:
:+1: makes sense to just exclude `FULLTEXT`